network/src/transport.rs

```diff
@@ -207,156 +218,948 @@ pub async fn perform_handshake<T: TSocket>(
+pub struct LibraNetTransport<TTransport> {
+    base_transport: TTransport,
+    ctxt: Arc<UpgradeContext>,
+    identity_pubkey: x25519::PublicKey,
+}
+
+impl<TTransport> LibraNetTransport<TTransport>
+where
+    TTransport: Transport<Error = io::Error>,
+    TTransport::Output: TSocket,
+    TTransport::Outbound: Send + 'static,
+    TTransport::Inbound: Send + 'static,
+    TTransport::Listener: Send + 'static,
+{
+    pub fn new(
+        base_transport: TTransport,
+        identity_key: x25519::PrivateKey,
+        trusted_peers: Option<Arc<RwLock<HashMap<PeerId, NetworkPublicKeys>>>>,
+        handshake_version: u8,
+        application_protocols: SupportedProtocols,
+    ) -> Self {
+        let mut own_handshake = HandshakeMsg::new();
+        own_handshake.add(SUPPORTED_MESSAGING_PROTOCOL, application_protocols);
+        let identity_pubkey = identity_key.public_key();
+
+        // Only use anti replay protection in mutual-auth scenarios. In theory,
+        // this is applicable everywhere; however, we would need to spend some
+        // time making this more sophisticated so it garbage collects old
+        // timestamps and doesn't use unbounded space. These are not problems in
+        // mutual-auth scenarios because we have a bounded set of trusted peers
+        // that rarely changes.
+        let anti_replay_timestamps = if trusted_peers.is_some() {
+            Some(Arc::new(RwLock::new(AntiReplayTimestamps::default())))
+        } else {
+            None
+        };
+
+        Self {
+            ctxt: Arc::new(UpgradeContext {
+                noise: NoiseWrapper::new(identity_key),
+                trusted_peers,
+                anti_replay_timestamps,
+                handshake_version,
+                own_handshake,
+            }),
+            base_transport,
+            identity_pubkey,
+        }
+    }
+
+    fn parse_dial_addr(
+        addr: &NetworkAddress,
+    ) -> io::Result<(NetworkAddress, x25519::PublicKey, u8)> {
+        use libra_network_address::Protocol::*;
+
+        let protos = addr.as_slice();
+
+        // parse out the base transport protocol(s), which we will just ignore
+        // and leave for the base_transport to actually parse and dial.
+        let (base_transport_protos, base_transport_suffix) = None
+            // TODO(philiphayes): protos[..X] is kinda hacky. `Transport` trait
+            // should handle this.
+            .or_else(|| parse_ip_tcp(protos).map(|x| (&protos[..2], x.1)))
+            .or_else(|| parse_dns_tcp(protos).map(|x| (&protos[..2], x.1)))
+            .or_else(|| parse_memory(protos).map(|x| (&protos[..1], x.1)))
+            .ok_or_else(|| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!(
+                        "Unexpected dialing network address: '{}', expected: \
+                         memory, ip+tcp, or dns+tcp",
+                        addr
+                    ),
```

Comment:
yeah, this feels weird, we should explicitly know which type we support by the builder we construct... otherwise we will potentially parse_memory in a ip network :O

Reply:
this is hacky because the `base_transport` doesn't have a `parse_dial_addr` or what not. ideally, you would let `base_transport` like parse out its bit, then parse out your piece, but `Transport` would need to change to support that. I would also have to figure out how the combinators would change to support this, which is explicitly what I wanted to avoid.